File: evaluation/common-csharp/Utils/DynamicBenchmark.cs
using BenchmarkDotNet.Attributes;
using System.Diagnostics;

namespace CommonCSharp.Utils
{
    /// <summary>
    /// BenchmarkDotNet benchmark class used by QueryExecutor.
    ///
    /// Mirrors Java's DynamicBenchmark: at runtime QueryExecutor sets
    /// BenchmarkHolder.Current to the query lambda, then calls
    /// BenchmarkRunner.Run() which drives the warmup and measurement iterations
    /// through this method. The warmup iteration is therefore recorded as the
    /// first entry in BenchmarkHolder.IterationResults; QueryExecutor strips it.
    ///
    /// [IterationSetup] forces GC before every iteration, matching JMH's
    /// shouldDoGC(true). Each invocation records elapsed time (Stopwatch) and
    /// memory allocated and appends them to BenchmarkHolder.IterationResults so
    /// the per-iteration detail table in the frontend continues to work.
    ///
    /// Memory uses GC.GetTotalAllocatedBytes (process-wide cumulative allocation),
    /// NOT GetAllocatedBytesForCurrentThread. The query is awaited, and an async
    /// MongoDB call allocates its result on the driver's I/O / continuation
    /// threads while the continuation may resume on a different thread than the
    /// one that captured allocBefore. A thread-local counter therefore compares
    /// two unrelated threads' totals and frequently yields 0 (negative delta
    /// clamped). The process-wide counter is immune to that thread hop. This is
    /// the async-C# analogue of Java's per-thread ThreadMXBean, which is valid
    /// there only because the Java queries are synchronous. It is safe because
    /// the orchestrator benchmarks one query at a time, so almost nothing else
    /// allocates during the measured window. Both counters are cumulative and
    /// unaffected by GC, so a collection cannot zero the measurement.
    ///
    /// Returning the result from [Benchmark] prevents the JIT from eliminating
    /// the call as dead code, matching JMH's Blackhole.consume().
    /// </summary>
    public class DynamicBenchmark
    {
        [IterationSetup]
        public void IterationSetup()
        {
            GC.Collect();
            GC.WaitForPendingFinalizers();
            GC.Collect();
        }

        [Benchmark]
        public async Task<int> Run()
        {
            long allocBefore = GC.GetTotalAllocatedBytes(precise: true);
            long t0 = Stopwatch.GetTimestamp();

            int size = await BenchmarkHolder.Current!();

            double elapsedMs = Stopwatch.GetElapsedTime(t0).TotalMilliseconds;
            long allocAfter = GC.GetTotalAllocatedBytes(precise: true);
            long delta = Math.Max(0L, allocAfter - allocBefore);

            BenchmarkHolder.LastSize = size;

            lock (BenchmarkHolder.IterationResults)
            {
                BenchmarkHolder.IterationResults.Add(new IterationResult
                {
                    elapsed = elapsedMs,
                    delta = delta,
                    result = size,
                    status = "success",
                    jfr = new JfrResult { totalAllocated = delta }
                });
            }

            return size;
        }
    }
}

File: evaluation/common/src/main/java/cz/cuni/mff/java/kurinna/common/utils/DynamicBenchmark.java
package cz.cuni.mff.java.kurinna.common.utils;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.infra.Blackhole;

import java.lang.management.ManagementFactory;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * JMH benchmark class used by QueryExecutor.
 *
 * The annotation processor generates the benchmark harness for this class at
 * compile time. At runtime, QueryExecutor sets SupplierHolder.current to the
 * query lambda, then calls Runner.run() which drives warmup and measurement
 * iterations through this method.
 *
 * Each invocation records elapsed time (nanoTime) and memory allocated
 * (ThreadMXBean) and appends them to SupplierHolder.iterationResults so the
 * per-iteration detail table in the frontend continues to work.
 */
@State(Scope.Thread)
@BenchmarkMode(Mode.SingleShotTime)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
public class DynamicBenchmark {

    private static final com.sun.management.ThreadMXBean THREAD_MX;

    static {
        java.lang.management.ThreadMXBean bean = ManagementFactory.getThreadMXBean();
        THREAD_MX = bean instanceof com.sun.management.ThreadMXBean ?
                (com.sun.management.ThreadMXBean) bean : null;
    }

    @Benchmark
    public void run(Blackhole bh) {
        long threadId = Thread.currentThread().threadId();

        long allocBefore = THREAD_MX != null ? THREAD_MX.getThreadAllocatedBytes(threadId) : 0L;
        long t0 = System.nanoTime();

        Collection<?> result = SupplierHolder.current.get();

        //==================================
        // Print result for testing purposes
        System.out.println("Query results:");
        result.stream()
                .limit(3)
                .forEach(System.out::println);

        System.out.println("Query size: " + result.size());
        //==================================

        long elapsedNs = System.nanoTime() - t0;
        long allocAfter = THREAD_MX != null ? THREAD_MX.getThreadAllocatedBytes(threadId) : 0L;
        long delta = Math.max(0L, allocAfter - allocBefore);

        // Consume result to prevent JIT dead-code elimination
        bh.consume(result);
        SupplierHolder.lastSize = result.size();

        // Build per-iteration entry matching the frontend's ServiceMetricType shape.
        // jfr sub-map keeps the same keys as before; only totalAllocated is populated
        // since JMH/ThreadMXBean does not provide the JFR-specific TLAB breakdown.
        Map<String, Object> jfr = new HashMap<>();
        jfr.put("totalAllocated", delta);
        jfr.put("gcCount", 0);
        jfr.put("heapUsedAvg", 0);
        jfr.put("allocatedInsideTLAB", 0);
        jfr.put("allocatedOutsideTLAB", 0);

        Map<String, Object> iterResult = new HashMap<>();
        iterResult.put("elapsed", elapsedNs / 1_000_000.0);
        iterResult.put("delta", delta);
        iterResult.put("result", result.size());
        iterResult.put("status", "success");
        iterResult.put("jfr", jfr);

        SupplierHolder.iterationResults.add(iterResult);
    }
}
